kotlin/workflow-runtime/src/main/java/com/squareup/workflow/internal/RealWorkflowHost.kt

```diff
@@ -0,0 +1,111 @@
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.produce
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.selects.select
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.coroutineContext
+
+/**
+ * @param initialState Allows unit tests to start the host from a given state, instead of calling
+ * [StatefulWorkflow.initialState].
+ */
+internal class RealWorkflowHost<InputT : Any, StateT : Any, OutputT : Any, RenderingT : Any>(
+  workflow: StatefulWorkflow<InputT, StateT, OutputT, RenderingT>,
+  context: CoroutineContext,
+  initialInput: InputT,
+  private val initialSnapshot: Snapshot?,
+  private val initialState: StateT? = null
+) : WorkflowHost<InputT, OutputT, RenderingT> {
+
+  private val scope = CoroutineScope(context)
+  private val inputs = Channel<InputT>(capacity = UNLIMITED)
+      .apply { offer(initialInput) }
+
+  override val updates: ReceiveChannel<Update<OutputT, RenderingT>> =
+    scope.produce(capacity = 0) {
+      runWorkflowTree(
+          workflow = workflow,
+          onUpdate = ::send
+      )
+    }
+
+  override fun setInput(input: InputT) {
+    inputs.offer(input)
+  }
+
+  /**
+   * Loops forever, or until the coroutine is cancelled, processing the workflow tree and emitting
+   * updates by calling [onUpdate].
+   */
+  private suspend fun runWorkflowTree(
+    workflow: StatefulWorkflow<InputT, StateT, OutputT, RenderingT>,
+    onUpdate: suspend (Update<OutputT, RenderingT>) -> Unit
+  ): Nothing {
+    var output: OutputT? = null
+    var input: InputT = inputs.receive()
+    var inputsClosed = false
+    val workflowNode = WorkflowNode(
+        id = workflow.id(),
+        workflow = workflow,
+        initialInput = input,
+        snapshot = initialSnapshot,
+        baseContext = coroutineContext,
+        initialState = initialState
+    )
+
+    try {
+      while (true) {
+        coroutineContext.ensureActive()
+
+        val rendering = workflowNode.compose(workflow, input)
+        val snapshot = workflowNode.snapshot(workflow)
+
+        onUpdate(Update(rendering, snapshot, output))
+
+        // Tick _might_ return an output, but if it returns null, it means the state or a child
+        // probably changed, so we should re-compose/snapshot and emit again.
+        output = select {
+          // While the inputs channel is still open, select on it so we can detect new inputs.
+          if (!inputsClosed) {
+            @Suppress("EXPERIMENTAL_API_USAGE")
+            inputs.onReceiveOrNull { newInput ->
+              if (newInput == null) {
+                inputsClosed = true
+              } else {
+                input = newInput
+              }
+              // No output.
```

Comment:
"No output. Returning from the select will go to the top of the loop to do another compose pass."